src/Services/QueuedJobService.php

```diff
@@ -1533,6 +1579,271 @@ private function addJobHandlersToLogger(LoggerInterface $logger, QueuedJob $job,
+                    $retryFalloffMultiplierVariance
+                );
+
+                // Apply multiplier to total delay, factor in number of retries
+                $delay *= pow($multiplier, $retryCount);
+                $delay = (int) floor($delay);
+            } else {
+                // Initial delay is missing - retry right away
+                $delay = 0;
+            }
+
+            $dateMessageSegment = 'immediately';
+
+            // Add a delay if configured, so we wouldn't retry the job immediately, but with some spread
+            // this lowers the chance of multiple jobs accessing the same data, thus preventing conflicts
+            if ($delay > 0) {
+                $scheduledTime = $now->getTimestamp() + $delay;
+                $jobDescriptor->StartAfter = $scheduledTime;
+
+                $formattedDate = DBDatetime::create_field('Datetime', $scheduledTime);
+                $dateMessageSegment = sprintf('after %s', $formattedDate->Rfc2822());
+            }
+
+            // Capture exception in the messages of the stuck job for better debug options
+            $message = sprintf('Automatic job retry - job will retry %s', $dateMessageSegment);
+
+            $this->copyDescriptorToJob($jobDescriptor, $job);
+
+            // Mark our retry attempt
+            $jobDescriptor->RetryCount += 1;
+
+            // Change the job status according to our map configuration
+            $jobDescriptor->JobStatus = $jobStatusMap[$jobDescriptor->JobStatus];
+
+            // Release the job lock, so it could be picked up again as well as expiry
+            $this->releaseJobLock($jobDescriptor);
+            $jobDescriptor->Expiry = null;
+
+            // This extension hook is useful not only to customise the message but also to integrate
+            // with another logging solution such as Graylog which provides the capability
+            // to see number of job retries over time as a visual chart
+            $this->extend('updateAutoRetryJobMessage', $jobDescriptor, $job, $message);
+
+            $job->addMessage($message);
+            $this->copyJobToDescriptor($job, $jobDescriptor);
+            $jobDescriptor->write();
+        }
+    }
+
+    /**
+     * Generate a randomised multiplier variance
+     */
+    protected function getJobRetryDelayMultiplier(float $multiplier, float $variance): float
+    {
+        // Variance is not in use or invalid
+        if (!$variance || $variance >= $multiplier) {
+            return $multiplier;
+        }
+
+        // Convert to integers for simpler calculation
+        $baseValue = (int) floor($multiplier * 100);
+        $offsetValue = (int) floor($variance * 100);
+
+        // This needs to be wrapped inside a separate non-private method so we can override it in
+        // situations such as tests as randomisation is not desirable in those cases
+        $multiplierKey = $this->getRandomValueFromRange($baseValue - $offsetValue, $baseValue + $offsetValue);
```

Comment:
```suggestion
        $multiplierKey = $this->getRandomValueFromRange(max(100, $baseValue - $offsetValue), $baseValue + $offsetValue);
```
I don't think we want multiplier to dip below `1` (which is `100` due to the multiplication above) - i.e. we should avoid _reducing_ the delay time.

The alternative would be to `return max(1, $multiplierKey / 100);` below - which is probably better actually.

Reply:
Looks like you might have missed this one

(edit: Adding a link since GitHub shows this comment on its own below the review: https://github.com/silverstripe/silverstripe-queuedjobs/pull/467#discussion_r2508494845)